src/libopensc/card-myeid.c

```diff
@@ -1185,6 +1254,144 @@ static int myeid_decipher(struct sc_card *card, const u8 * crgram,
 }
 
 
+static int myeid_wrap_key(struct sc_card *card, u8 *out, size_t outlen)
+{
+	struct sc_context *ctx;
+	struct sc_apdu apdu;
+	u8 rbuf[SC_MAX_APDU_BUFFER_SIZE];
+	int r;
+
+	assert(card != NULL);
+	ctx = card->ctx;
+	LOG_FUNC_CALLED(ctx);
+
+	/* INS: 0x2A  PERFORM SECURITY OPERATION
+	   P1:  0x84  Resp: Return a cryptogram
+	 * P2:  0x00  The data field is absent */
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0x2A, 0x84, 0x00);
+	apdu.resp = rbuf;
+	apdu.resplen = sizeof(rbuf);
+	apdu.le = 0;
+	apdu.lc = 0;
+
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(ctx, r, "APDU transmit failed");
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(ctx, r, "wrap key failed");
+
+	if (apdu.resplen <= outlen && out != NULL)
+		memcpy(out, apdu.resp, apdu.resplen);
+
+	LOG_FUNC_RETURN(ctx, apdu.resplen);
+}
+
+static int myeid_unwrap_key(struct sc_card *card, const u8 *crgram, size_t crgram_len)
+{
+	int r;
+	u8 out[512];
+	size_t outlen = 512;
+	struct sc_apdu apdu;
+	u8 rbuf[SC_MAX_APDU_BUFFER_SIZE];
+	u8 sbuf[SC_MAX_APDU_BUFFER_SIZE];
+	u8 p2 = 0x86;
+	myeid_private_data_t* priv;
+
+	LOG_FUNC_CALLED(card->ctx);
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_NORMAL);
+
+	assert(card != NULL && crgram != NULL && out != NULL);
+
+	if (crgram_len > 256)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+
+	if (card->drv_data)
+	{
+		priv = (myeid_private_data_t*) card->drv_data;
+
+		if (priv->sec_env)
+		{
+			if (priv->sec_env->algorithm == SC_ALGORITHM_AES ||
+				priv->sec_env->algorithm == SC_ALGORITHM_3DES ||
+				priv->sec_env->algorithm == SC_ALGORITHM_DES)
+						p2 = 0x84;		/* Set correct P2 for symmetric crypto */
+		}
+	}
+
+	/* INS: 0x2A  PERFORM SECURITY OPERATION
+	    * P1:  0x00  Do not expect response - the deciphered data will be placed into the target key EF.
+	    * P2:  0x86  Cmd: Padding indicator byte followed by cryptogram
+		* P2:  0x84  Cmd: AES/3DES Cryptogram (plain value encoded in BER-TLV DO, but not including SM DOs) */
+	sc_format_apdu(card, &apdu,
+		(crgram_len < 256) ? SC_APDU_CASE_4_SHORT : SC_APDU_CASE_3_SHORT,
+		0x2A, 0x00, p2);
+
+	apdu.resp = rbuf;
+	apdu.resplen = sizeof(rbuf);
+	apdu.le = crgram_len;
+
+	if (crgram_len == 256 && p2 == 0x86)
```

Comment:
I'll do some refactoring for this function. The code has originally been copied from myeid_decipher, where 256 byte (2048 bit RSA) input data is split into two APDUs, and we set LE only for the second APDU, because we don't expect return data from the first one. 

When we are doing unwrapping, actually we don't return any data from the card, as the decrypted data is placed into a file on card, so we need only SC_APDU_CASE_3_SHORT, it seems. 

I removed some other useless code from this function too. I'll have to test this first, and commit the changes then.